### PR TITLE
implemented installroot option

### PR DESCRIPTION
**Pull request by Alexander198961**
##### SUMMARY

added installroot option to zypper module

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

module
##### ANSIBLE VERSION

```
ansible 2.4.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.9 (default, Dec 21 2014, 11:02:59) [GCC
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/packaging/os/zypper.py
Comment (robinro) [inline lib/ansible/modules/packaging/os/zypper.py]: What about `update` and `patch`? Do those work with `--root` as well?

File: test/integration/targets/zypper/tasks/zypper.yml
Comment (robinro) [inline test/integration/targets/zypper/tasks/zypper.yml]: This URL will break in a few days/weeks/months since old versions are not kept in the repo.
Do you need the .rpm file, or can you just use zypper to talk to the repos directly?

Possible better alternatives:
* use package without version
* use self-build package (see existing tests)

**Review by robinro — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/packaging/os/zypper.py
Comment (Alexander198961) [reply lib/ansible/modules/packaging/os/zypper.py]: it seems update and patch   don't have this option if I understand zypper docs correct

File: test/integration/targets/zypper/tasks/zypper.yml
Comment (Alexander198961) [reply test/integration/targets/zypper/tasks/zypper.yml]: I will try with download package from suse repo and install package without version

**Comment by Alexander198961** [comment]
does this ok now @robinro  ? I have modified test a little bit it will download package using zypper -d (download option)

File: lib/ansible/modules/packaging/os/zypper.py
Comment (robinro) [reply lib/ansible/modules/packaging/os/zypper.py]: Since `--root` comes before the subcommand it should work on all subcommands. How would you update a package in a different installroot otherwise?

File: test/integration/targets/zypper/tasks/zypper.yml
Comment (robinro) [inline test/integration/targets/zypper/tasks/zypper.yml]: Don't use `/tmp`, but the `{{output_dir | expanduser}}` to not write files outside the test environment.

Comment (robinro) [reply test/integration/targets/zypper/tasks/zypper.yml]: There is already a valid rpm built inside the test cases. It would be easier to use that file instead of downloading/moving an external file.

**Review by robinro — CHANGES_REQUESTED** [review]
Overall I'm not sure it's the best solution to add an argument for `installroot`.
Maybe we can instead add `extra_args_precommand`?
There already is `extra_args` for custom arguments after the install command and one could use a similar variable for arguments before the command. I don't think adding all options explicitly is the right solution, since the list is fairly long and most are rarely used (including `--root`).

**Comment by Alexander198961** [comment]
@robinro  added extra_args_precommand  and updated test

**Comment by robinro** [comment]
@Alexander198961 Thanks for the changes. If this "extra_args_precommand" setup works for you I think it's nicer than the first version.
There are some unrelated formatting changes and copy-paste leftovers. Please have a look at the diff and clean it up.

Before this gets merged I'd like some feedback by the core team on the argument name. `extra_args_precommand` is telling in my view, but it's also fairly long and maybe there is a better choice based on the experience in other modules.

File: lib/ansible/modules/packaging/os/zypper.py
Comment (robinro) [inline lib/ansible/modules/packaging/os/zypper.py]: double space and space before dot

File: lib/ansible/modules/packaging/os/zypper.py
Comment (robinro) [inline lib/ansible/modules/packaging/os/zypper.py]: state: present
installed is only an alias and should not be used in examples

File: lib/ansible/modules/packaging/os/zypper.py
Comment (robinro) [inline lib/ansible/modules/packaging/os/zypper.py]: ?

File: test/integration/targets/zypper/tasks/zypper.yml
Comment (robinro) [inline test/integration/targets/zypper/tasks/zypper.yml]: double space

File: test/integration/targets/zypper/tasks/zypper.yml
Comment (robinro) [inline test/integration/targets/zypper/tasks/zypper.yml]: double space

File: test/integration/targets/zypper/tasks/zypper.yml
Comment (robinro) [inline test/integration/targets/zypper/tasks/zypper.yml]: use stat module instead

**Review by robinro — CHANGES_REQUESTED** [review]
Some more cleanup changes.

File: lib/ansible/modules/packaging/os/zypper.py
Comment (Alexander198961) [reply lib/ansible/modules/packaging/os/zypper.py]: @robinro if i amn't mistaken in this case code tries to find something like this zypper search --match-exact apache-rpm-macros-control-20151110-8.1.x86_64.rpm .  so not sure if zypper search will work in this way so tried to exclude rpm packages from search .  function get_installed_state

Comment (robinro) [reply lib/ansible/modules/packaging/os/zypper.py]: `urls` are not checked for presence, but always passed along to zypper. See the rest of this function.

**Comment by robinro** [comment]
Looks good now. Thanks @Alexander198961 for your work!
shipit

**Comment by Alexander198961** [comment]
should we close this pr ?

**Review by robinro — APPROVED** [review]

**Comment by robinro** [comment]
@ansible core
shipit

**Comment by andytom** [comment]
shipit

**Comment by gundalow** [comment]
rebuild_merge

**Comment by robinro** [comment]
@Alexander198961 This PR is not mergable at the moment. Please rebase this branch and clean up conflicts
http://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html

To get it merged quickly, ping one of the core members on IRC #ansible-devel

**Comment by gundalow** [comment]
Moved to https://github.com/ansible/ansible/pull/33613/ reviews welcome so we can get this into 2.5

**Pull request closed without merging**
